modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/IgniteMdSelectivity.java

```diff
@@ -109,7 +149,612 @@ public Double getSelectivity(IgniteSortedIndexSpool rel, RelMetadataQuery mq, Re
+                return new BigDecimal(value.getFloat());
+
+            case Value.DATE:
+                return new BigDecimal(value.getDate().getTime());
+
+            case Value.TIME:
+                return new BigDecimal(value.getTime().getTime());
+
+            case Value.TIMESTAMP:
+                return new BigDecimal(value.getTimestamp().getTime());
+
+            case Value.BYTES:
+                BigInteger bigInteger = new BigInteger(1, value.getBytes());
+                return new BigDecimal(bigInteger);
+
+            case Value.STRING:
+            case Value.STRING_FIXED:
+            case Value.STRING_IGNORECASE:
+            case Value.ARRAY:
+            case Value.JAVA_OBJECT:
+            case Value.GEOMETRY:
+                return null;
+
+            case Value.UUID:
+                BigInteger bigInt = new BigInteger(1, value.getBytes());
+                return new BigDecimal(bigInt);
+
+            default:
+                throw new IllegalStateException("Unsupported H2 type: " + value.getType());
+        }
+    }
+
+    /**
+     * Predicate based selectivity for table. Estimate condition on each column taking in comparison it's statistics.
+     *
+     * @param rel Original rel node to fallback calculation by.
+     * @param tbl Underlying IgniteTable.
+     * @param mq RelMetadataQuery
+     * @param predicate Predicate to estimate selectivity by.
+     * @return Selectivity.
+     */
+    private double getTablePredicateBasedSelectivity(RelNode rel, IgniteTable tbl, RelMetadataQuery mq, RexNode predicate) {
+        if (tbl == null)
+            return super.getSelectivity(rel, mq, predicate);
+
+        double sel = 1.0;
+
+        Map<ColumnStatistics, Boolean> addNotNull = new HashMap<>();
+
+        for (RexNode pred : RelOptUtil.conjunctions(predicate)) {
+            ColumnStatistics colStat = getColStat(rel, mq, tbl, pred);
+
+            if (pred.getKind() == SqlKind.IS_NULL)
+                sel *= estimateIsNullSelectivity(rel, mq, tbl, pred);
+            else if (pred.getKind() == SqlKind.IS_NOT_NULL) {
+                if (colStat != null)
+                    addNotNull.put(colStat, Boolean.FALSE);
+
+                sel *= estimateIsNotNullSelectivity(rel, mq, tbl, pred);
+            } else if (pred.isA(SqlKind.EQUALS)) {
+                if (colStat != null) {
+                    Boolean colNotNull = addNotNull.get(colStat);
+
+                    addNotNull.put(colStat, (colNotNull == null) || colNotNull);
+                }
+
+                sel *= estimateEqualsSelectivity(rel, mq, tbl, pred);
+            } else if (pred.isA(SqlKind.COMPARISON)) {
+                if (colStat != null) {
+                    Boolean colNotNull = addNotNull.get(colStat);
+
+                    addNotNull.put(colStat, (colNotNull == null) || colNotNull);
+                }
+
+                sel *= estimateComparisonSelectivity(rel, mq, tbl, pred);
+            } else
+                sel *= .25;
```

Comment:
looks like every OR operator will fallback to this branch. Is it possible to analyze and disjunctions too?